### PR TITLE
Defined Request/response headers that proxy to pass to/from a FHIR server

### PR DESCRIPTION
**Pull request by vivekmittal07**
## Description of what I changed
Added support for headers defined in https://www.hl7.org/fhir/http.html#ops and https://www.hl7.org/fhir/async.html

If there are multiple values for the same header then only 1 was being sent to FHIR server. Caught this issue in test cases.

Fixes https://github.com/google/fhir-access-proxy/issues/44

## E2E test

TESTED:
Ran the server locally to verify the request and response headers. Allowed response headers are now being returned in the curl request. Ran the server in debug mode to verify if the Allowed request headers are passed to FHIR if caller is sending those headers.

## Checklist: I completed these to help reviewers :)

- [x] I have read and will follow the
      [review process](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/blob/master/doc/review_process.md).
- [x] I am familiar with Google Style Guides for the language I have coded in.

  No? Please take some time and review
  [Java](https://google.github.io/styleguide/javaguide.html) and
  [Python](https://google.github.io/styleguide/pyguide.html) style guides. Note,
  when in conflict, OpenMRS style guide overrules.

- [x] My IDE is configured to follow the Google
      [**code styles**](https://google.github.io/styleguide/).

  No? Unsure? ->
  [configure your IDE](https://github.com/google/google-java-format).

- [x] I have **added tests** to cover my changes. (If you refactored existing
      code that was well tested you do not have to add tests)
- [x] I ran `mvn clean package` right before creating this pull request and
      added all formatting changes to my commit.
- [x] All new and existing **tests passed**.
- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: server/src/main/java/com/google/fhir/proxy/HttpFhirClient.java [hunk 1]
```diff
@@ -43,7 +43,38 @@ public abstract class HttpFhirClient {
...
+          "fhirVersion",
+          "if-match",
+          "if-none-match",
+          "if-modified-since",
+          "if-unmodified-since",
+          "if-range",
+          "x-request-id",
+          "x-correlation-id",
+          "x-forwarded-for",
+          "x-forwarded-host");
```
Comment (bashir2) [inline server/src/main/java/com/google/fhir/proxy/HttpFhirClient.java:77]: Question: What is the benefit of passing this? Have you tried to see if this makes any difference on how HAPI or GCP FHIR servers respond to queries? (ditto for `x-forwarded-for`)
What I am wondering is (i) what is the scenario where we receive these and whether we should keep them and (ii) if they make a difference, maybe we should _set_ them when querying the FHIR servers (for example I am thinking if setting those may solve [this problem](https://github.com/google/fhir-access-proxy/blob/cd1e01e502b3aa4798f29f7e72f5909479417601/server/src/main/java/com/google/fhir/proxy/BearerAuthorizationInterceptor.java#L345)).

File: server/src/main/java/com/google/fhir/proxy/HttpFhirClient.java [hunk 2]
```diff
@@ -43,7 +43,38 @@ public abstract class HttpFhirClient {
...
+          "x-progress",
+          "x-request-id",
+          "x-correlation-id");
+
+  // The list of incoming header names to keep for forwarding request to the FHIR server; use lower
+  // case only.
+  // We should NOT copy Content-Length as this is automatically set by the RequestBuilder when
+  // setting content Entity; otherwise we will get a ClientProtocolException.
+  static final Set<String> REQUEST_HEADERS_TO_KEEP =
+      Sets.newHashSet(
```
Comment (bashir2) [inline server/src/main/java/com/google/fhir/proxy/HttpFhirClient.java:63]: Should we include `accept` in this set too?

File: server/src/main/java/com/google/fhir/proxy/HttpFhirClient.java [hunk 3]
```diff
@@ -43,7 +43,38 @@ public abstract class HttpFhirClient {
 
   // The list of header names to keep in a response sent from the proxy; use lower case only.
   // Note we don't copy content-length/type because we may modify the response.
-  static final Set<String> HEADERS_TO_KEEP = Sets.newHashSet("last-modified", "date");
+  static final Set<String> RESPONSE_HEADERS_TO_KEEP =
```
Comment (bashir2) [inline server/src/main/java/com/google/fhir/proxy/HttpFhirClient.java:46]: Please consider adding a reference to the FHIR documentation from which these HEADERs come from (for both sets). I think something like what you have in the PR description is good enough (doesn't need to be detailed for each).

File: server/src/main/java/com/google/fhir/proxy/HttpFhirClient.java [hunk 4]
```diff
@@ -43,7 +43,38 @@ public abstract class HttpFhirClient {
...
+  // setting content Entity; otherwise we will get a ClientProtocolException.
+  static final Set<String> REQUEST_HEADERS_TO_KEEP =
+      Sets.newHashSet(
+          "content-type",
+          "last-modified",
+          "etag",
+          "prefer",
+          "fhirVersion",
+          "if-match",
+          "if-none-match",
```
Comment (bashir2) [inline server/src/main/java/com/google/fhir/proxy/HttpFhirClient.java:70]: Should we also include `if-none-exist`? In general, how can we make sure that these two sets are exhaustive?

**Review by bashir2 — COMMENTED** [review]
Thanks @vivekmittal07 for fixing this. I just had some minor comments/questions.

**Comment by vivekmittal07** [comment]
I have created a new PR from a forked repo - http://go/gh/google/fhir-access-proxy/pull/55
I don't have access to push a new commit to this PR anymore.

**Comment by bashir2** [comment]
> I have created a new PR from a forked repo - http://go/gh/google/fhir-access-proxy/pull/55 I don't have access to push a new commit to this PR anymore.

Sure, let's move to the new PR. Maybe just finish the discussion threads here then close this PR (so no new commits here).

**Pull request closed without merging**